### ISSUE TITLE
Direct sequence of events

### ISSUE THREAD

**Issue by Thylossus** [jruizgit/rules#25]
Hi,

first of all, thank you for developing and maintaining this great project. What I will explain next is probably no technical issue. I'm using the TypeScript API and want to detect three subsequent "error" events in a stream of events.

Each event has a type property and I tried to define a rule that is similar to the rule defined for https://github.com/jruizgit/rules/blob/master/docs/js/reference.md#correlated-sequence

This rule should trigger the action if three events of type error are posted right after each other, e.g. "info ok error error error", but not if there are other events in between, e.g. "info error ok error error".

This is my code:

```
import * as d from 'durable';
const { m, s, c } = d;
let counter = 0;

function createId(): string {
    counter += 1;
    return Date.now() + ' ' + counter;
}

const set: any = d.ruleset('test', 
    {
        whenAll: [
            c['id0'] = m['type'].eq('error'),
            c['id1'] = m['type'].eq('error'),
            c['id2'] = m['type'].eq('error'),
        ],
        run: (c) => {
            console.log('TRIGGERED: 3 error events in a row');
            Object.keys(c).forEach(key => {
                if (key.match(/^id/)) {
                    console.log(c[key]);
                }
            });
        }
    }
);

set.whenStart((host) => {
    console.log('inserting events');
    const sid: number = 2;
    for (let i = 0; i < 25; i++) {
        setTimeout(() => {
            const type: string = ['ok', 'info', 'error'][Math.floor(Math.random() * 3)];
            const id: string = createId();
            console.log(id, type);
            host.post('test', {
                id,
                sid,
                type,
            });
        }, Math.random() * 5000);
    }
});

console.log('run...');
d.runAll([{ host: '192.168.99.100', port: 6379 }]);
```

Unfortunately, the action is triggered whenever an error event is observed. For example, the program produced the following output: 

```
run...
inserting events
Listening on 5000
1469713486428 1 ok
1469713486724 2 ok
1469713487548 3 ok
1469713487624 4 info
1469713487763 5 ok
1469713488117 6 ok
1469713488208 7 ok
1469713488401 8 ok
1469713488415 9 ok
1469713488732 10 ok
1469713488860 11 ok
1469713488883 12 info
1469713488985 13 error
TRIGGERED: 3 error events in a row
{ id: '1469713488985 13', type: 'error', sid: 2 }
{ id: '1469713488985 13', type: 'error', sid: 2 }
{ id: '1469713488985 13', type: 'error', sid: 2 }
1469713489349 14 info
1469713489415 15 info
1469713489647 16 error
1469713489798 17 ok
1469713489851 18 ok
TRIGGERED: 3 error events in a row
{ id: '1469713489647 16', type: 'error', sid: 2 }
{ id: '1469713489647 16', type: 'error', sid: 2 }
{ id: '1469713489647 16', type: 'error', sid: 2 }
1469713489990 19 info
1469713490027 20 info
1469713490261 21 ok
1469713490309 22 error
TRIGGERED: 3 error events in a row
{ id: '1469713490309 22', type: 'error', sid: 2 }
{ id: '1469713490309 22', type: 'error', sid: 2 }
{ id: '1469713490309 22', type: 'error', sid: 2 }
1469713490483 23 ok
1469713490678 24 ok
1469713490907 25 error
TRIGGERED: 3 error events in a row
{ id: '1469713490907 25', type: 'error', sid: 2 }
{ id: '1469713490907 25', type: 'error', sid: 2 }
{ id: '1469713490907 25', type: 'error', sid: 2 }
```

**Comment by jruizgit** [comment]
Hi, thanks for posting the question. The reason why each event triggers the rule is because an error event alway satisfies the three clauses. You can use the count rule decorator, however it will not guarantee consecutiveness. 

```typescript
const set: any = d.ruleset('test', 
    {
        count: 3, 
        whenAll: [
            m['type'].eq('error'),
        ],
        run: (c) => {
            console.log('TRIGGERED: 3 error events in a row' + JSON.stringify(c.m));
            console.log(c.m.length);
            c.m.forEach(row => {
                console.log('Error -> ' + row['id']);
            });
        }
    }
);
```

If you can guarantee a strict auto-increment sequence when generating events you can compare ids.

**Comment by Thylossus** [comment]
Hi, thank you for your response. I've already tried the count rule decorator, but as you mentioned it does not guarantee consecutiveness. Thus, I will opt for comparing ids.

**Comment by jruizgit** [comment]
Hi, I have been playing with this scenario a little more.

First, you might have noticed there is a bug when writing a rule (in typescript or js) that compares ids. I just fixed the bug and published a new build. 

Guaranteeing auto-increment ids or accurate timestamps when originating events in a distributed system is a difficult problem to solve. Trying to reason about consecutive messages comparing consecutive Ids or timestamps is not advisable. The good news is: there is a better way.

From a purely declarative (no side-effects) perspective, it is possible to write a set of rules to decide what to do with three consecutive messages. It is important to define the outcome of each combination.

```typescript
const set: any = d.ruleset('test', 
    {
        whenAll: [
            c['id0'] = m['type'].eq('error'),
            c['id1'] = m['type'].eq('error').and(m['id'].neq(c['id0']['id'])),
            c['id2'] = m['type'].eq('error').and(m['id'].neq(c['id0']['id'])).and(m['id'].neq(c['id1']['id'])),
        ],
        run: (c) => {
            console.log('TRIGGERED: 3 error events in a row');
        },
    },
    {
        whenAll: [
            c['id0'] = m['type'].neq('error'),
            c['id1'] = m['type'].eq('error').and(m['id'].neq(c['id0']['id'])),
            c['id2'] = m['type'].eq('error').and(m['id'].neq(c['id0']['id'])).and(m['id'].neq(c['id1']['id'])),
        ],
        run: (c) => {
            console.log('TRIGGERED: 1 success and 2 error events in a row');
        },
    },
    {
        whenAll: [
            c['id0'] = m['type'].neq('error'),
            c['id1'] = m['type'].neq('error').and(m['id'].neq(c['id0']['id'])),
            c['id2'] = m['type'].eq('error').and(m['id'].neq(c['id0']['id'])).and(m['id'].neq(c['id1']['id'])),
        ],
        run: (c) => {
            console.log('TRIGGERED: 2 success and 1 error event in a row');
        },
    },
    {
        whenAll: [
            c['id0'] = m['type'].neq('error'),
            c['id1'] = m['type'].neq('error').and(m['id'].neq(c['id0']['id'])),
            c['id2'] = m['type'].neq('error').and(m['id'].neq(c['id0']['id'])).and(m['id'].neq(c['id1']['id'])),
        ],
        run: (c) => {
            console.log('TRIGGERED: 3 success events in a row');
        },
    }
);
```
While the ruleset above will detect three consecutive error messages in a tumbling window, you might want to detect three consecutive error messages in a sliding window. 

Being practical, it is possible to use the ruleset state (or context):

```typescript
const set: any = d.ruleset('test', 
    {
        whenAll: [
            s['errors'].gte(3),
        ],
        run: (c) => {
            console.log('3 consecutive errors detected');
            c['s']['errors'] = 0;       
        },
    },
    {
        whenAll: [
            m['type'].ex(),
        ],
        run: (c) => {
            if (c['m']['type'] == 'error') {
                c['s']['errors'] = isNaN(c['s']['errors']) ? 0: c['s']['errors'] + 1;
            } else {
                c['s']['errors'] = 0;
            }
        },
    }
);
```
The ruleset above will detect three consecutive errors in a sliding window. There is a performance penalty: an action is triggered for every event, while in the first example an action is triggered for every 3 events.

**Comment by Thylossus** [comment]
Wow! Thank you for investigating this for me and providing the solutions. These solutions are definitely easier to scale than the ID approach.

**Comment by sandeeps2** [comment]
How would you incorporate time component to this? Correlate a sequence of events E1 and E2 within an interval of 't' seconds starting from the arrival of E1. Any events before E1 should be discarded.

I tried to do this via the following statechart with a manual reset timer. It fires even if E2 comes before E1.

```javascript
var d = require('durable');

d.statechart('risk', function() {

    start: {
        to: 'waitTrigger'
        // will start a manual reset timer
        run: {
            console.log('Start ' + new Date());
            // startTimer('VelocityTimer', 15, true);
        }
    }

    waitTrigger: {
        to: 'meter'
        whenAll: {
          m.event == 'Trigger'
        }
        run: {
            console.log('Got trigger, reset and start timer ' + new Date());
            // s.trigger = 'Trigger'
            resetTimer('VelocityTimer');
            startTimer('VelocityTimer', 15, true);
        }
    }

    meter: {
        to: 'waitTrigger'
        whenAll: { 
            m.amount > 100 /*&& s.trigger == 'Trigger'*/
            timeout('VelocityTimer')
        }
        cap: 100
        run: {
            console.log('Velocity: ' + m.length + ' events in 15 seconds with context ' +  new Date()  + ' ' /*+ s.trigger*/);
            deleteState();
            // s.trigger = 'Reset'
        }  

        to: 'waitTrigger'
        whenAll: {
            timeout('VelocityTimer')
        }
        run: {
            console.log('Timeout ' + s.trigger + ' ' + new Date());
            deleteState();
            // s.trigger = 'Reset'
        }
    }

    whenStart: {
/*
        post('risk', { event: 'Trigger' }); 
        post('risk', { amount: 200 }); 
        post('risk', { amount: 150 }); 
        post('risk', { amount: 50 }); 
        post('risk', { amount: 500 }); 
        post('risk', { amount: 600 }); 
*/
    }
});
d.runAll();
```
The following event should not be allowed if there is no preceding `{"event":"Trigger"}` but the outcome is 0 and a subsequent Trigger event satisfies the rule, which is strange.
```
curl -H "content-type: application/json" -X POST http://localhost:5000/risk/events -d '{"amount":150}'
{
  "outcome": 0
}
```
